proxy/server/server.go

```diff
@@ -357,3 +377,144 @@ func newScopeRouter(scope string, router http.Handler) (http.Handler, error) {
 	}
 	return scopedRouter, nil
 }
+
+// Returns a GetCertificate function for this host based on AutoCert settings,
+// by generating an autocert manager for the host.
+// It optionally will staple the OCSP response to the cert when it is generated,
+// if OCSP stapling is enabled for the host.
+func certFuncForHost(hostCfg engine.Host, autoCertCache autocert.Cache, s stapler.Stapler) (getCertificateFunc, error) {
+	ac := hostCfg.Settings.AutoCert
+
+	// Each host gets its own Autocert Manager - this allows individual
+	// certs to use different autocert authorities, as well as auth keys
+	autoCertMgr := &autocert.Manager{
+		Prompt:     autocert.AcceptTOS,
+		Cache:      autoCertCache,
+		HostPolicy: autocert.HostWhitelist(hostCfg.Name),
+		Email:      ac.Email,
+	}
+
+	if ac.RenewBefore > 0 {
+		autoCertMgr.RenewBefore = ac.RenewBefore
+	}
+
+	// if either directory or key are non-empty, we need to generate
+	// a custom ACME client to override either.
+	if ac.Key != "" || ac.DirectoryURL != "" {
+
+		// If DirectoryURL is empty, the default Let's Encrypt URL will be picked.
+		autoCertMgr.Client = &acme.Client{
+			DirectoryURL: ac.DirectoryURL,
+		}
+
+		// If Key is non-empty, then decode it as RSA or EC which are the only two keys
+		// we support. Go's crypto library doesn't support a generic function to provide back
+		// a private key interface.
+		if ac.Key != "" {
+			block, _ := pem.Decode([]byte(ac.Key))
+			if block == nil {
+				return nil, fmt.Errorf("Autocert Key PEM Block for Host %s is invalid.", hostCfg.Name)
+			} else if block.Type == "RSA PRIVATE KEY" {
+				rsaPrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+				if err == nil {
+					autoCertMgr.Client.Key = rsaPrivateKey
+				} else {
+					log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an RSA Private Key: %v", block.Type, hostCfg.Name, err)
+				}
+			} else if block.Type == "EC PRIVATE KEY" {
+				ecPrivateKey, err := x509.ParseECPrivateKey(block.Bytes)
+				if err == nil {
+					autoCertMgr.Client.Key = ecPrivateKey
+				} else {
+					log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an EC Private Key: %v", block.Type, hostCfg.Name, err)
+				}
+			} else {
+				return nil, fmt.Errorf("AutoCert Private Key for Host %s is of unrecognized type: %s. Supported types"+
+					"are RSA PRIVATE KEY and EC PRIVATE KEY.", hostCfg.Name, block.Type)
+			}
+		}
+	}
+
+	// Wrap the GetCert for this host, so we can generate and staple
+	// an optional OCSP response when requested.
+	stapledGetCert := func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		keyPair, err := autoCertMgr.GetCertificate(info)
+
+		if hostCfg.Settings.OCSP.Enabled {
```

Comment:
Can't we just call `ocspStapleToCert` here? The code seems identical.

Reply:
Aha. I am fixing that. The reason for passing the GetCertFunc into stapler is when the stapler updates the OCSP payload, the cert might have changed. I added an extra param to ocspStapleToCert to pass that function through.